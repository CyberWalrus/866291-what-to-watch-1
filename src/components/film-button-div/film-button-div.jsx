import React, {PureComponent, Fragment} from "react";
import PropTypes from "prop-types";
import {Link} from "react-router-dom";
import {connect} from "react-redux";
import RoutePath from "../../routes.js";
import {getError} from "../../store/data/selectors.js";
import {getAuthorizationStatus} from "../../store/user/selectors.js";
import {Operation} from "../../store/data/data.js";
import {ActionCreator} from "../../store/filter/filter.js";
import {FavoriteStatus} from "../../mock/constants.js";

class FilmButtonDiv extends PureComponent {
  constructor(props) {
    super(props);
    this._handlOnClickFavorite = this._handlOnClickFavorite.bind(this);
    this._handlOnClickPlay = this._handlOnClickPlay.bind(this);
  }
  _handlOnClickPlay() {
    const {id} = this.props;
    this.props.onVideoScreenOpen(id);
  }
  _handlOnClickFavorite() {
    const {id, isFavorite} = this.props;
    const status = isFavorite ? FavoriteStatus.REMOVE : FavoriteStatus.ADD;
    this.props.sendFavorite(status, id);
  }
  render() {
    const {
      id,
      isFavorite,
      isAuthorizationRequired,
      isShowReview
    } = this.props;
    return (
      <div className="movie-card__buttons">
        <button
          className="btn btn--play movie-card__button"
          type="button"
          onClick={this._handlOnClickPlay}
        >
          <svg viewBox="0 0 19 19" width="19" height="19">
            <use xlinkHref="#play-s" />
          </svg>
          <span>Play</span>
        </button>
        {isAuthorizationRequired ? (
          <button
            className="btn btn--list movie-card__button"
            type="button"
            onClick={this._handlOnClickFavorite}
          >
            <svg viewBox="0 0 19 20" width="19" height="20">
              <use xlinkHref={isFavorite ? `#remove` : `#add`} />
            </svg>
            <span>My list</span>
          </button>
        ) : (
          <Link
            to={RoutePath.LOGIN}
            className="btn btn--list movie-card__button"
          >
            <svg viewBox="0 0 19 20" width="19" height="20">
              <use xlinkHref={`#add`} />
            </svg>
            <span>My list</span>
          </Link>
        )}

        {isShowReview ? (
          <Link
            to={RoutePath.ADD_REVIEW.replace(`:id`, id)}
            className="btn movie-card__button"
          >
            Add review
          </Link>
        ) : (
          <Fragment />
        )}
      </div>
    );
  }
}
FilmButtonDiv.propTypes = {
  id: PropTypes.number.isRequired,
  isFavorite: PropTypes.bool.isRequired,
  isAuthorizationRequired: PropTypes.bool.isRequired,
  sendFavorite: PropTypes.func.isRequired,
  onVideoScreenOpen: PropTypes.func.isRequired,
  isShowReview: PropTypes.bool
};
const mapStateToProps = (state, ownProps) =>
  Object.assign({}, ownProps, {
    errorMessage: getError(state),
    isAuthorizationRequired: getAuthorizationStatus(state)
  });
const mapDispatchToProps = (dispatch) => ({
  sendFavorite: (status, filmId) =>
    dispatch(Operation.sendFavorite(status, filmId)),
  onVideoScreenOpen: (filmId) =>
    dispatch(ActionCreator.setPlayFilmId(filmId))
});
export {FilmButtonDiv};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(FilmButtonDiv);
