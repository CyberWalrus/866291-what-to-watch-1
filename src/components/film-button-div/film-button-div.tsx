import * as React from "react";
import {PureComponent, Fragment} from "react";
import {Link} from "react-router-dom";
import {connect} from "react-redux";
import RoutePath, {routeToReview} from "../../routes.js";
import {getAuthorizationStatus} from "../../store/user/selectors.js";
import {Operation} from "../../store/data/data.js";
import {ActionCreator} from "../../store/filter/filter.js";
import {FavoriteStatus} from "../../constants.js";

interface Props {
  id: number,
  isFavorite: boolean,
  isAuthorizationRequired: boolean,
  isShowReview: boolean,
  onSendFavorite: (status: number, id: number) => void,
  onVideoScreenOpen: (id: number) => void,
}
class FilmButtonDiv extends PureComponent<Props, null> {
  constructor(props) {
    super(props);
    this._handleClickFavorite = this._handleClickFavorite.bind(this);
    this._handleClickPlay = this._handleClickPlay.bind(this);
  }
  _handleClickPlay() {
    const {id} = this.props;
    this.props.onVideoScreenOpen(id);
  }
  _handleClickFavorite() {
    const {id, isFavorite} = this.props;
    const status = isFavorite ? FavoriteStatus.REMOVE : FavoriteStatus.ADD;
    this.props.onSendFavorite(status, id);
  }
  render() {
    const {id, isFavorite, isAuthorizationRequired, isShowReview} = this.props;
    return (
      <div className="movie-card__buttons">
        <button
          className="btn btn--play movie-card__button"
          type="button"
          onClick={this._handleClickPlay}
        >
          <svg viewBox="0 0 19 19" width="19" height="19">
            <use xlinkHref="#play-s" />
          </svg>
          <span>Play</span>
        </button>
        {isAuthorizationRequired ? (
          <button
            className="btn btn--list movie-card__button"
            type="button"
            onClick={this._handleClickFavorite}
          >
            <svg viewBox="0 0 19 20" width="19" height="20">
              <use xlinkHref={isFavorite ? `#remove` : `#add`} />
            </svg>
            <span>My list</span>
          </button>
        ) : (
          <Link
            to={RoutePath.LOGIN}
            className="btn btn--list movie-card__button"
          >
            <svg viewBox="0 0 19 20" width="19" height="20">
              <use xlinkHref={`#add`} />
            </svg>
            <span>My list</span>
          </Link>
        )}

        {isShowReview ? (
          <Link
            to={routeToReview(id)}
            className="btn movie-card__button"
          >
            Add review
          </Link>
        ) : (
          <Fragment />
        )}
      </div>
    );
  }
}

const mapStateToProps = (state, ownProps) =>
  Object.assign({}, ownProps, {
    isAuthorizationRequired: getAuthorizationStatus(state)
  });
const mapDispatchToProps = (dispatch) => ({
  onSendFavorite: (status, filmId) =>
    dispatch(Operation.sendFavorite(status, filmId)),
  onVideoScreenOpen: (filmId) => dispatch(ActionCreator.setPlayFilmId(filmId))
});

export {FilmButtonDiv};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(FilmButtonDiv);
