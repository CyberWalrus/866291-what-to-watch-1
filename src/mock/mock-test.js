const FILMS = [
  {
    bgColor: `#977461`,
    description: `In 1954, a U.S. Marshal investigates the disappearance of a murderer, who escaped from a hospital for the criminally insane.`,
    director: `Martin Scorsese`,
    genre: `Thriller`,
    id: 1,
    isFavorite: false,
    pageUrl: `Shutter Island`,
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    srcPreviewVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    rating: `8,1`,
    ratingLevel: `Very good`,
    released: 2010,
    runTime: 138,
    scoresCount: 1002557,
    srcBgImage: `https://es31-server.appspot.com/wtw/static/film/background/Shutter_Island.jpg`,
    srcPosterImage: `https://es31-server.appspot.com/wtw/static/film/poster/Shutter_Island.jpg`,
    srcPreviewImage: `https://es31-server.appspot.com/wtw/static/film/preview/shutter-island.jpg`,
    starrings: [(`Leonardo DiCaprio`, `Emily Mortimer`, `Mark Ruffalo`)],
    title: `Shutter Island`
  },
  {
    bgColor: `#F1E9CE`,
    description: `Macbeth, the Thane of Glamis, receives a prophecy from a trio of witches that one day he will become King of Scotland. Consumed by ambition and spurred to action by his wife, Macbeth murders his king and takes the throne for himself.`,
    director: `Justin Kurzel`,
    genre: `Drama`,
    id: 2,
    isFavorite: false,
    pageUrl: `Macbeth`,
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    srcPreviewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    rating: `6,6`,
    ratingLevel: `Good`,
    released: 2015,
    runTime: 113,
    scoresCount: 48798,
    srcBgImage: `https://es31-server.appspot.com/wtw/static/film/background/Macbeth.jpg`,
    srcPosterImage: `https://es31-server.appspot.com/wtw/static/film/poster/Macbeth.jpg`,
    srcPreviewImage: `https://es31-server.appspot.com/wtw/static/film/preview/macbeth.jpg`,
    starrings: [(`Michael Fassbender`, `Marion Cotillard`, `Jack Madigan`)],
    title: `Macbeth`
  }
];
const FILM = {
  bgColor: `#977461`,
  description: `In 1954, a U.S. Marshal investigates the disappearance of a murderer, who escaped from a hospital for the criminally insane.`,
  director: `Martin Scorsese`,
  genre: `Thriller`,
  id: 1,
  isFavorite: false,
  pageUrl: `Shutter Island`,
  srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
  srcPreviewVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
  rating: `8,1`,
  ratingLevel: `Very good`,
  released: 2010,
  runTime: 138,
  scoresCount: 1002557,
  srcBgImage: `https://es31-server.appspot.com/wtw/static/film/background/Shutter_Island.jpg`,
  srcPosterImage: `https://es31-server.appspot.com/wtw/static/film/poster/Shutter_Island.jpg`,
  srcPreviewImage: `https://es31-server.appspot.com/wtw/static/film/preview/shutter-island.jpg`,
  starrings: [(`Leonardo DiCaprio`, `Emily Mortimer`, `Mark Ruffalo`)],
  title: `Shutter Island`
};
const REVIEWS = [
  {
    comment: `test test test test test test test test test test test test test`,
    date: `2019-05-11`,
    dateHTML: `May 11, 2019`,
    id: 1,
    rating: `8,1`,
    userId: 1,
    userName: `Walrus`
  },
  {
    comment: `test test test test test test test test test test test test test`,
    date: `2019-05-11`,
    dateHTML: `May 11, 2019`,
    id: 1,
    rating: `8,1`,
    userId: 1,
    userName: `Walrus`
  }
];

const GENRES = [
  `All genres`,
  `Thriller`,
  `Drama`,
  `Adventure`,
  `Comedy`,
  `Crime`,
  `Action`,
  `Fantasy`
];
export {FILMS, FILM, GENRES, REVIEWS};
