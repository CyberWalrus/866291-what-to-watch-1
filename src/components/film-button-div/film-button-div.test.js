import React from "react";
import renderer from "react-test-renderer";
import {BrowserRouter} from "react-router-dom";
import {FilmButtonDiv} from "./film-button-div.jsx";
import {FILM} from "../../mock/mock-test.js";

it(`FilmButtonDiv correctly renders after relaunch`, () => {
  const handleClick = jest.fn();
  const tree = renderer
    .create(
        <BrowserRouter>
          <FilmButtonDiv
            id={FILM.id}
            isFavorite={false}
            onSendFavorite={handleClick}
            isAuthorizationRequired={false}
            onVideoScreenOpen={handleClick}
          />
        </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
