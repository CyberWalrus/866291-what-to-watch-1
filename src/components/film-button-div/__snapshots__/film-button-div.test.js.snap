// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmButtonDiv correctly renders after relaunch 1`] = `
<div
  className="movie-card__buttons"
>
  <button
    className="btn btn--play movie-card__button"
    onClick={[Function]}
    type="button"
  >
    <svg
      height="19"
      viewBox="0 0 19 19"
      width="19"
    >
      <use
        xlinkHref="#play-s"
      />
    </svg>
    <span>
      Play
    </span>
  </button>
  <a
    className="btn btn--list movie-card__button"
    href="/login"
    onClick={[Function]}
  >
    <svg
      height="20"
      viewBox="0 0 19 20"
      width="19"
    >
      <use
        xlinkHref="#add"
      />
    </svg>
    <span>
      My list
    </span>
  </a>
</div>
`;
